popup_offer/src/SpecialProductsDiscount.php

```diff
@@ -1,101 +0,0 @@
-<?php
-namespace Drupal\popup_offer;
-
-use Drupal\commerce_order\Adjustment;
-use Drupal\commerce_product\Entity\ProductVariation;
-use Drupal\Core\Entity\EntityRepositoryInterface;
-use Drupal\Core\Entity\EntityTypeManagerInterface;
-use Drupal\Core\Form\FormBuilderInterface;
-use Drupal\Core\Form\FormState;
-
-class SpecialProductsDiscount {
-
-  /**
-   * The entity type manager.
-   *
-   * @var \Drupal\Core\Entity\EntityTypeManagerInterface
-   */
-  protected $entityTypeManager;
-
-  /**
-   * The form builder.
-   *
-   * @var \Drupal\Core\Form\FormBuilderInterface
-   */
-  protected $formBuilder;
-
-  /**
-   * The entity repository.
-   *
-   * @var \Drupal\Core\Entity\EntityRepositoryInterface
-   */
-  protected $entityRepository;
-
-  /**
-   * Constructs a new CartLazyBuilders object.
-   *
-   * @param \Drupal\Core\Entity\EntityTypeManagerInterface $entity_type_manager
-   *   The entity type manager.
-   * @param \Drupal\Core\Form\FormBuilderInterface $form_builder
-   *   The form builder.
-   * @param \Drupal\Core\Entity\EntityRepositoryInterface $entity_repository
-   *   The entity repository.
-   */
-  public function __construct(EntityTypeManagerInterface $entity_type_manager, FormBuilderInterface $form_builder, EntityRepositoryInterface $entity_repository) {
-    $this->entityTypeManager = $entity_type_manager;
-    $this->formBuilder = $form_builder;
-    $this->entityRepository = $entity_repository;
-  }
-
-  /**
-   * Builds the add to cart form.
-   *
-   * @param string $product_id
-   *   The product ID.
-   * @param string $view_mode
-   *   The view mode used to render the product.
-   * @param bool $combine
-   *   TRUE to combine order items containing the same product variation.
-   * @param string $langcode
-   *   The langcode for the language that should be used in form.
-   *
-   * @return array
-   *   A renderable array containing the cart form.
-   */
-  public function addToCartForm($product, $view_mode, $combine, $langcode, $discount = null) {
-    /** @var \Drupal\commerce_order\OrderItemStorageInterface $order_item_storage */
-    $order_item_storage = $this->entityTypeManager->getStorage('commerce_order_item');
-    // Load Product for current language.
-    $product = $this->entityRepository->getTranslationFromContext($product, $langcode);
```

Comment:
I'm not sure we need $this->entityRepository. we can load entity using e.g. $this->entityTypeManager->getStorage('commerce_order_item')->load($entity_id)

Reply:
Most of the code I took from commerce module, just adjusted with some functionality I needed for applying discount.